cmd/bosun/sched/host.go

```diff
@@ -0,0 +1,809 @@
+			}
+			idPlus++
+			fstatus, timestamp, err := s.Search.GetLast("hw.ps", ts.String(), false)
+			status := "Bad"
+			if fstatus == 0 {
+				status = "Ok"
+			}
+			current, currentTimestamp, _ := s.Search.GetLast("hw.chassis.current.reading", opentsdb.TagSet{"host": host.Name, "id": fmt.Sprintf("PS%v", idPlus)}.String(), false)
+			volts, voltsTimestamp, _ := s.Search.GetLast("hw.chassis.volts.reading", opentsdb.TagSet{"host": host.Name, "name": fmt.Sprintf("PS%v_Voltage_%v", idPlus, idPlus)}.String(), false)
+			ps := &PowerSupply{}
+			if err == nil && timestamp > 0 {
+				ps.Status = status
+				ps.StatusLastUpdated = timestamp
+				ps.Amps = current
+				ps.AmpsLastUpdated = currentTimestamp
+				ps.Volts = volts
+				ps.VoltsLastUpdated = voltsTimestamp
+				host.Hardware.PowerSupplies[id] = ps
+			}
+			for _, m := range hostMetadata {
+				if m.Time.Before(time.Now().Add(-timeFilterAge)) {
+					continue
+				}
+				if !m.Tags.Equal(ts) {
+					continue
+				}
+				switch val := m.Value.(type) {
+				case string:
+					switch m.Name {
+					case "psMeta":
+						err = json.Unmarshal([]byte(val), &ps)
+						if err != nil {
+							slog.Errorf("error unmarshalling power supply meta for host %s, while generating host api: %s", host.Name, err)
+						} else {
+							host.Hardware.PowerSupplies[id] = ps
+						}
+					}
+				}
+			}
+		}
+		for _, ts := range hwBatteriesTags {
+			if ts["host"] != host.Name {
+				continue
+			}
+			id, ok := ts["id"]
+			if !ok {
+				slog.Errorf("couldn't find battery id tag %s for host %s", host.Name, id)
+				continue
+			}
+			fstatus, timestamp, err := s.Search.GetLast("hw.storage.battery", ts.String(), false)
+			status := "Bad"
+			if fstatus == 0 {
+				status = "Ok"
+			}
+			if err == nil && timestamp > 0 {
+				host.Hardware.Storage.Batteries[id] = &Battery{
+					Status:            status,
+					StatusLastUpdated: timestamp,
+				}
+			}
+		}
+		for _, ts := range hwBoardPowerTags {
+			if ts["host"] != host.Name {
+				continue
+			}
+			fstatus, timestamp, err := s.Search.GetLast("hw.chassis.power.reading", ts.String(), false)
+			if err == nil && timestamp > 0 {
+				host.Hardware.BoardPowerReading = &BoardPowerReading{
+					Watts:            int64(fstatus),
+					WattsLastUpdated: timestamp,
+				}
+			}
+		}
+		for _, ts := range hwPhysicalDiskTags {
+			if ts["host"] != host.Name {
+				continue
+			}
+			id, ok := ts["id"]
+			if !ok {
+				return nil, fmt.Errorf("couldn't find physical disk id tag for host %s", host.Name)
+			}
+			fstatus, timestamp, err := s.Search.GetLast("hw.storage.pdisk", ts.String(), false)
+			status := "Bad"
+			if fstatus == 0 {
+				status = "Ok"
+			}
+			pd := &PhysicalDisk{}
+			if err == nil && timestamp > 0 {
+				pd.Status = status
+				pd.StatusLastUpdated = timestamp
+				host.Hardware.Storage.PhysicalDisks[id] = pd
+			}
+			for _, m := range hostMetadata {
+				if m.Time.Before(time.Now().Add(-timeFilterAge)) {
+					continue
+				}
```

Comment:
In the scenario, where would you recommend I check for the existence of the host in that particular tagset result?